x/profiles/module_ibc.go

```diff
@@ -149,28 +184,93 @@ func (am AppModule) OnRecvPacket(
+// handleOracleRequestPacketData tries handling a OracleResponsePacketData packet.
+// Returns ErrInvalidPacketData if the given packet data is not of such type.
+func (am AppModule) handleOracleRequestPacketData(
+	ctx sdk.Context, packet channeltypes.Packet,
+) (channeltypes.Acknowledgement, error) {
+	var data oracletypes.OracleResponsePacketData
+	if err := oracletypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		return channeltypes.Acknowledgement{}, sdkerrors.Wrapf(types.ErrInvalidPacketData, "%T", packet)
+	}
+
+	acknowledgement := channeltypes.NewResultAcknowledgement([]byte{byte(1)})
+
+	err := am.keeper.OnRecvApplicationLinkPacketData(ctx, data)
 	if err != nil {
-		return nil, []byte{}, sdkerrors.Wrap(sdkerrors.ErrInvalidType, err.Error())
+		acknowledgement = channeltypes.NewErrorAcknowledgement(err.Error())
 	}
 
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypePacket,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.ModuleName),
+			sdk.NewAttribute(types.AttributeKeyClientID, data.ClientID),
+			sdk.NewAttribute(types.AttributeKeyRequestID, fmt.Sprintf("%d", data.RequestID)),
+			sdk.NewAttribute(types.AttributeKeyResolveStatus, data.ResolveStatus.String()),
+			sdk.NewAttribute(types.AttributeKeyAckSuccess, fmt.Sprintf("%t", err != nil)),
+		),
+	)
+
 	// NOTE: acknowledgement will be written synchronously during IBC handler execution.
-	return &sdk.Result{
-		Events: ctx.EventManager().Events().ToABCIEvents(),
-	}, ackBytes, nil
+	return acknowledgement, nil
 }
 
 // OnAcknowledgementPacket implements the IBCModule interface
 func (am AppModule) OnAcknowledgementPacket(
 	ctx sdk.Context,
-	modulePacket channeltypes.Packet,
+	packet channeltypes.Packet,
 	acknowledgement []byte,
 ) (*sdk.Result, error) {
+	var ack channeltypes.Acknowledgement
+	if err := types.AminoCdc.UnmarshalJSON(acknowledgement, &ack); err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest,
+			"cannot unmarshal oracle packet acknowledgement: %v", err)
+	}
+
+	var data oracletypes.OracleRequestPacketData
+	if err := oracletypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest,
+			"cannot unmarshal oracle request packet data: %s", err.Error())
+	}
+
+	if err := am.keeper.OnOracleRequestAcknowledgementPacket(ctx, data, ack); err != nil {
+		return nil, err
+	}
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypePacket,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.ModuleName),
+			sdk.NewAttribute(types.AttributeKeyClientID, data.ClientID),
+			sdk.NewAttribute(types.AttributeKeyAck, fmt.Sprintf("%v", ack)),
+		),
+	)
+
+	switch resp := ack.Response.(type) {
+	case *channeltypes.Acknowledgement_Result:
+		ctx.EventManager().EmitEvent(
+			sdk.NewEvent(
+				types.EventTypePacket,
+				sdk.NewAttribute(types.AttributeKeyAckSuccess, string(resp.Result)),
+			),
+		)
+	case *channeltypes.Acknowledgement_Error:
```

Comment:
This case can not be reached cause the error acknowledgment would be returned by `OnOracleRequestAcknowledgementPacket`.
The error should not be returned early if  we would like to emit event with error attribute here.

Reply:
This is not actually true though. Inside `OnOracleRequestAcknowledgementPacket`, the error is returned only if the storing of the updated link errors: 

```golang
return k.SaveApplicationLink(ctx, user, link)
```

Inside such method, the acknowledgement error is only used to update the link state properly:

```golang
case *channeltypes.Acknowledgement_Error:
	// The acknowledgment failed on the receiving chain.
	// Update the state to ERROR and the result to an error one
	link.State = types.AppLinkStateVerificationError
	link.Result = types.NewErrorResult(res.Error)
```

Reply:
@RiccardoM You are right. I missed that thing.